### ISSUE TITLE
(Inline) documentation

### ISSUE THREAD

**Issue by fonfon** [elastic/elasticsearch-dsl-py#372]
I was trying to reconstruct a query that I already have working with django-haystack (combining various OR-filters), and came across a [stackoverflow comment](http://stackoverflow.com/a/32417230/1138355) where somebody used `from elasticsearch_dsl import F`.

So I had a look at the source code but couldn't find any F. Was it removed? The [release notes](http://elasticsearch-dsl.readthedocs.org/en/latest/Changelog.html) don't tell anything about it (warning bells start ringing).

Looking at the code I found SF[](https://github.com/elastic/elasticsearch-dsl-py/blob/master/elasticsearch_dsl/function.py#L3). Is it SearchFunction? ScoreFunction? SearchFilter? It's a public function without docstring which is not mentioned in the documentation (warning bells are getting louder).

Long story short: I will query ES directly now because this seems easier than trying to understand the internals of elasticsearch-dsl-py before being able to query. Better inline or readthedocs querying documentation (examples for combining filters) would have helped me.

**Comment by HonzaKral** [comment]
`F` stands for filter and those have been deprecated in `2.0` release and is mentioned in the Changelog (though it doesn't mention `F` explicitly, just added that). The problem is that readthedocs builds have been failing, I apologize for that and it has been fixed.

Documentation with examples for combining filters can be found in both README and on read the docs, could you please specify what exactly is missing for you.

The entire function score mechanism is not documented, that is of course wrong, but there were other priorities and function score can work without `SF` (`ScoreFunction`) just as well and many people use it just by following the mechanism for creating queries described in the docs.

Thanks for the report!

**Comment by fonfon** [comment]
Thank you for your quick response!

TLDR: I managed to construct my query using Q -- I think more filter examples (and a correct answer to the SO-question) would be good.

Details:
The stackoverflow-answer pointed me at the wrong direction. I tried to combine various filters using  `search.filter("or", list_of_filters)` but this failed because `filter` isn't intended to be used like that anymore I guess.

The README shows how to filter for one specific property:
`.filter("term", category="search")`
but I wanted to OR-combine various such filters like described [at stackoverflow](http://stackoverflow.com/questions/28564303/how-do-i-create-an-or-filter-using-elasticsearch-dsl-py) (but deprecated because of F), so something like (I'm inventing the `filter_or` here):
```
search.filter("term", category="category1")\
.filter_or(Q("term", category="category2") & Q("term", color="blue"))
```

I understand (only) now that I can construct one large Q by combining Qs like `myQ = Q("term", category="category2") & Q("term", color="blue")` and apply it like that:
`search.query('bool', filter=[myQ])`.

At the point of reading, for me Q was Query but I looked for Filter so I didn't spend enough attention at the Q at http://elasticsearch-dsl.readthedocs.org/en/latest/search_dsl.html#filters. First I wanted to filter and just then to query, so I searched for the F and off I went.

**Comment by HonzaKral** [comment]
The answer was correct at the time, but then new version of elasticsearch and this library came out. Unfortunately there is absolutely no way for us to keep the answers on SO up to date over time.

For `or` logic you can just use the `|` operator, which is described in the documentation exactly http://elasticsearch-dsl.readthedocs.org/en/latest/search_dsl.html#query-combination Alternatively you can just create a `bool` query with `should` clause yourself (which is what happens under the covers with `|`): `Q('bool', should=[Q(...), Q(...)])`

Filters no longer exist in elasticsearch itself, they have been replaced by queries - https://www.elastic.co/guide/en/elasticsearch/reference/2.2/query-filter-context.html

Unfortunately there is no way we can duplicate the entire documentation for elasticsearch itself in this library. You need to know how your version of elasticsearch works, then this library will hopefully save you some time in doing what you wish to do.

Because this library doesn't try to abstract away the details and options that elasticsearch has, it is very heavily dependent on functionality and development of elasticsearch itself. This was a very deliberate decision where we didn't want to hide away features and allow people to use the full power of elasticsearch.

**Comment by fonfon** [comment]
I just included the link to SO to show how I originally failed at constructing the query (and I posted an answer there now).

I personally like abstraction and the idea of an 'ES release-independent' higher-level API but I understand the disadvantages. Django-haystack was powerful enough for my querying purposes (but surprisingly not flexible enough at handling results) -- for now I'll continue using elasticsearch-dsl-py, thanks for your clarifications and the doc updates!
